src/grep.rs

```diff
@@ -197,13 +203,16 @@ fn grep_offset(
 /// Run bgrep with the given options, outputting to the given `StdoutLock`.
 /// Error detail may be outputted to stderr.
 /// Returns whether there was a match.
-fn run_file(
-  stdout: &mut io::StdoutLock,
+fn run_file<O: Write, P: Into<PathBuf>, B: AsMut<Vec<u8>>>(
```

Comment:
I'm not aware of the benefit of using `P: Into<PathBuf>` here instead of `PathBuf` itself. Could you elaborate please?

Reply:
In using `P: Into<PathBuf>`, I just want to mean that the argument is something that can be converted into `PathBuf` (e.g. a `& PathBuf`). So later, I can convert 
https://github.com/gahag/bgrep/blob/858ce10e53506c034f0223f269a7c686fc510928/src/grep.rs#L293-L297

into

```Rust
for file in files.iter() {
  match run_file(stdout, &options, &pattern, file, buffer) {
```

which I think "more lightweight" since we don't need clone a vector from the boxed slice `files`: we just iterate over it (and then `file` should be a reference: we don't own elements of the slice).

Reply:
I see. I think there was a misunderstanding here. The vector was not cloned, but consumed. The benefit of consuming the vector is that it will only be iterated once, and each value will be used and dropped immediately. When using a non-owned iterator, the array has to be iterated twice: one iteration to use each value, and one to free each value when the vector is dropped.

For this reason, I believe the code as it was is more performant. I'm aware that the speedup here would be minimal, as this is not really the processor intensive part of the code, but still I'd like to have it the most efficient way.

Therefore, I think we should keep just the `O: Write` and the `B: AsMut<Vec<u8>>>` changes.

Are you ok with that? Please let me know you if you have any further consideration.

Reply:
You're correct and many thanks for showing me that `files.to_vec()` is to own the data, I didn't know that :(

I didn't understand that your intention is to consume the vector at this point, so consuming it as soon as possible is also a good choice for me.